### ISSUE TITLE
Incorrect/confusing error message with multiple bounds of a single associated-type parametrized trait

### ISSUE THREAD

**Issue by syntacticsugarglider** [rust-lang/rust#70514]
```rust
#![allow(unreachable_code)]

pub trait Super<P> {
    type Assoc;
}

pub trait A<P>: Super<P> {}

pub trait B<P>: Super<P> {
    fn join(&mut self, handle: Self::Assoc) -> ();
}

pub trait Trait<T> {
    type Assoc;
}

// Removing B<<C as Trait<T>>::Assoc> bound allows this to infer and resolve
fn bug<T, U, C: Trait<T> + Trait<U> + B<<C as Trait<T>>::Assoc> + A<<C as Trait<T>>::Assoc> + B<<C as Trait<U>>::Assoc> + A<<C as Trait<U>>::Assoc>>() {
    let ctx: &mut C = panic!();
    let a: <C as Super<<C as Trait<U>>::Assoc>>::Assoc = panic!();
    
    // Removing this line allows it to resolve
    ctx.join(a);
}
```

I expect that this would compile given that the type of `a` is fully specified and only one implementation of `B` can satisfy the requirement imposed. Aside from that, the error message produced here is on `let ctx`, specifically "consider giving `ctx` the explicit type `&mut C`, where the type parameter `C` is specified". An error is in fact reported on `join`, 
```
cannot infer type for type parameter `C`

note: cannot resolve <C as Super<_>>::Assoc == <C as Super<<C as Trait<U>>::Assoc>>::Assoc
```
but in the real-world codebase from which this originated the error was reported in an earlier use of `ctx` in the function, not the one which caused the issue, and it would be reported on the first occurrence of a use of `C` by receiver in the function regardless of the nature of that use.

Fully qualifying the call i.e.
```rust
B::<<C as Trait<U>>::Assoc>::join(ctx, a);
```
also resolves and the program compiles successfully, my concern here is around the exceedingly confusing misplaced error message which sometimes has a span in unrelated source.

[playground reproduction](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a0fb72d5443c71262f275bc54b199f9e)

**Comment by steffahn** [comment]
I’m not sure this should compile. I would however have expected an error like the one that you will get after replacing the call with `C::join(ctx,a);`:
```
   Compiling playground v0.0.1 (/playground)
error[E0282]: type annotations needed
  --> src/lib.rs:31:5
   |
31 |     C::join(ctx,a);
   |     ^^^^^^^ cannot infer type for type parameter `P` declared on the trait `B`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0282`.
error: could not compile `playground`.

To learn more, run the command again with --verbose.
```

The current error, suggesting that the type `C` is not known and should be specified, for sure seems incorrect.

@rustbot modify labels to A-diagnostics, D-incorrect, and T-compiler.

**Comment by estebank** [comment]
Current output:

```
error[E0284]: type annotations needed
  --> src/lib.rs:23:9
   |
23 |     ctx.join(a);
   |         ^^^^
   |
   = note: cannot satisfy `<C as Super<_>>::Assoc == <C as Super<<C as Trait<U>>::Assoc>>::Assoc`
help: try using a fully qualified path to specify the expected types
   |
23 -     ctx.join(a);
23 +     <C as B<P>>::join(ctx, a);
   |
```
Following the suggestions you get

```
error[E0412]: cannot find type `P` in this scope
  --> src/lib.rs:23:13
   |
 7 | pub trait A<P>: Super<P> {}
   | ------------------------ similarly named trait `A` defined here
...
23 |     <C as B<P>>::join(ctx, a);
   |             ^
   |
help: a trait with a similar name exists
   |
23 -     <C as B<P>>::join(ctx, a);
23 +     <C as B<A>>::join(ctx, a);
   |
help: you might be missing a type parameter
   |
18 | fn bug<T, U, C: Trait<T> + Trait<U> + B<<C as Trait<T>>::Assoc> + A<<C as Trait<T>>::Assoc> + B<<C as Trait<U>>::Assoc> + A<<C as Trait<U>>::Assoc>, P>() {
   |                                                                                                                                                    +++
```
which then leads to other errors if you blindly follow the `P` to `A` replacement.
